kernel/src/device/tdxguest/mod.rs

```diff
@@ -109,12 +143,101 @@ fn handle_get_report(arg: usize) -> Result<i32> {
         return Err(err.into());
     }
 
-    let tdx_report_vaddr = arg + TDX_REPORTDATA_LEN;
-    let mut generated_report = vec![0u8; TDX_REPORT_LEN];
+    let mut generated_report = Box::new([0u8; TDX_REPORT_LEN]);
     dma_coherent
-        .read_bytes(1024, &mut generated_report)
+        .read_bytes(1024, generated_report.as_mut())
         .unwrap();
-    let report_slice: &[u8] = &generated_report;
-    user_space.write_bytes(tdx_report_vaddr, &mut VmReader::from(report_slice))?;
+
+    Ok(generated_report)
+}
+
+fn handle_get_report(arg: usize) -> Result<i32> {
+    let current_task = ostd::task::Task::current().unwrap();
+    let user_space = CurrentUserSpace::new(current_task.as_thread_local().unwrap());
+    let user_request: TdxReportRequest = user_space.read_val(arg)?;
+
+    let report = tdx_get_report(&user_request.report_data)?;
+
+    let tdx_report_vaddr = arg + TDX_REPORTDATA_LEN;
+    user_space.write_bytes(tdx_report_vaddr, &mut VmReader::from(report.as_ref()))?;
+
     Ok(0)
 }
+
+pub fn tdx_get_quote(inblob: &[u8]) -> Result<Box<[u8]>> {
+    const GET_QUOTE_IN_FLIGHT: u64 = 0xFFFF_FFFF_FFFF_FFFF;
+    const GET_QUOTE_BUF_SIZE: usize = 8 * 1024;
+
+    let report = tdx_get_report(inblob)?;
+    let entry = alloc_quote_entry(GET_QUOTE_BUF_SIZE);
+
+    entry
+        .buf
+        .write_bytes(offset_of!(TdxQuoteHdr, version), &1u64.to_le_bytes())?;
+    entry
+        .buf
+        .write_bytes(offset_of!(TdxQuoteHdr, status), &0u64.to_le_bytes())?;
+    entry.buf.write_bytes(
+        offset_of!(TdxQuoteHdr, in_len),
+        &(TDX_REPORT_LEN as u32).to_le_bytes(),
+    )?;
+    entry
+        .buf
+        .write_bytes(offset_of!(TdxQuoteHdr, out_len), &0u32.to_le_bytes())?;
+    entry.buf.write_bytes(size_of::<TdxQuoteHdr>(), &report)?;
+
+    if let Err(err) = get_quote(
+        (entry.buf.paddr() as u64) | SHARED_MASK,
+        entry.buf_len as u64,
+    ) {
+        error!("[kernel] get quote error: {:?}", err);
+        return Err(err.into());
+    }
+
+    let mut quote_buffer = [0u8; size_of::<TdxQuoteHdr>()];
+    let mut quote_hdr: TdxQuoteHdr;
+
+    // Poll for the quote to be ready.
+    loop {
+        entry.buf.read_bytes(0, &mut quote_buffer)?;
+        quote_hdr = parse_quote_header(&quote_buffer);
+        if quote_hdr.status != GET_QUOTE_IN_FLIGHT {
```

Comment:
Why not read the `status` field directly? Use` SafePtr` to read status directly.